### ISSUE TITLE
Stream large file to AsyncApp (application/octet-stream)

### ISSUE THREAD

**Issue by alfechner** [spec-first/connexion#1877]
### Description

I would like to stream a large file to my `AsyncApp`. My app is supposed to proxy the incoming bytes to another service. The goal is to reduce the memory footprint.

My spec looks like this:

```yaml
openapi: 3.0.3
info:
  title: LIQUID Dataset Service
  version: 1.0.0
paths:
  /foo:
    post:
      x-openapi-router-controller: my_controller
      operationId: my_controller_function
      requestBody:
        required: true
        content:
          application/octet-stream:
            schema:
              type: string
              format: binary
```

I receive the bytes of my file in `my_controller_function` but already fully loaded. Means, the entire file is in memory already.

Is there any way to get hand on the stream in the controller before fully transferring the entire content?

- `Python 3.12.1`
- `Version: 3.0.5`

**Comment by RobbeSneyders** [comment]
Hi @alfechner,

I think you have two options:
- Use a different content type like `multipart/form-data` instead.
  ```yaml
  requestBody:
  content:
    multipart/form-data:
      schema:
        type: object
        properties:
          file:
            type: string
            format: binary
  ```
  and accept the `file` parameter in your handler function:
  ```python
  def post(file):
      ...
  ```
  This will give you a [`starlette.UploadFile`](https://www.starlette.io/requests/#request-files) object.
- Access the body via the request instead:
  ```python
  from connexion import request

  def post():
      stream = request.stream()
  ```
  Check the [starlette docs](https://www.starlette.io/requests/#body) for more info on how to access the request.

  It's important to not accept the `body` parameter in your handler function, since that triggers the materialization of the stream by connexion.

Agree that this could be more clear in the docs.

**Comment by alfechner** [comment]
Hi @RobbeSneyders, really appreciate your helpful input here.

The `multipart/form-data` didn't work for me. I looked into the code and I believe that [starlette's form parser reads the entire stream into memory](https://github.com/encode/starlette/blob/74ccb961b3c1b1871aa7ed70a81dd3000e0194da/starlette/formparsers.py#L104) already before passing the [starlette.UploadFile](https://www.starlette.io/requests/#request-files) to the controller. I found [this comment](https://github.com/tiangolo/fastapi/issues/2578#issuecomment-752334996) confirming my finding.

I made the stream option work:

```python
async def test_stream():
    request = Request(scope, receive)
    async for chunk in request.stream():
       ....
```

Just a note: When passing the body parameter to `test_stream`, the `stream()` function runs into a infinitive loop. The `_stream_consumed` property is not set correctly but the stream is empty. I assume this is somehow related to the copying of the Request object.

**Comment by RobbeSneyders** [comment]
Ok, thanks for diving into this and reporting back.

Any reason you're creating a new request instead of importing the one we make available via `from connexion import request`?

**Comment by alfechner** [comment]
No, you're right. It works the exact same way with `from connexion.context import request`. The code above is a leftover from my attempt to work around the `infinitive loop` issue I mentioned. But it's not changing anything at all.

**Comment by RobbeSneyders** [comment]
Is the `infinite loop` issue you mention the same as described in https://github.com/spec-first/connexion/issues/1812?

**Comment by alfechner** [comment]
Looks like it, yes 👍

**Comment by alfechner** [comment]
I wanted to have another look at the `multipart/form-data`. I have a request body with (1) a file and (2) some metadata for that file:

```yaml
requestBody:
  required: true
  content:
    multipart/form-data:
      schema:
        type: object
        properties:
          file:
            type: string
            format: binary
          metadata:
            type: object
            additionalProperties: true
      encoding:
        metadata:
          contentType: application/json
```

I wanted to see if I can modify starlette's `MultiPartParser` in a way that I can intercept the stream while reading in the bytes of the file and stream it to the other server. The metadata would only be available after reading the entire stream (and sending it of) which would be fine in my case.

I tried to run the original `MultiPartParser` on the request to get started:

```python
from connexion.context import request
from starlette.formparsers import MultiPartParser

async def test_form_data():
    form_data = MultiPartParser(headers=request.headers, stream=request.stream()).parse()
```

I'm running into an `infinite loop` again.

Opposed to the `application/octet-stream` example there is a validator registered for body type `multipart/form-data`. This means the [validation takes place](https://github.com/spec-first/connexion/blob/211bdb03f6b20f7dc90fe0037e57cd142c48deb5/connexion/middleware/request_validation.py#L120-L126) before invoking the controller and "materializes" the stream.

In order to disable the validation I unset the body validator for `multipart/form-data` via the `validator_map` setting and verified that the verification is skipped.

Even tough there is no validation anymore I'm hanging in the `infinite loop` and I found the parameter [decorator to consume the stream](https://github.com/spec-first/connexion/blob/211bdb03f6b20f7dc90fe0037e57cd142c48deb5/connexion/decorators/parameter.py#L109) when loading the files. 

Would it make sense to add a ` _maybe_get_files` method like [_maybe_get_body](https://github.com/spec-first/connexion/blob/211bdb03f6b20f7dc90fe0037e57cd142c48deb5/connexion/decorators/parameter.py#L109)? What do you think?
